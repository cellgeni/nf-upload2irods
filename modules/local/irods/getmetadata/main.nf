process IRODS_GETMETADATA {
    tag "Getting metadata for ${irodspath}"

    input:
    tuple val(meta), val(irodspath)

    output:
    tuple val(meta), path("irods_metadata.csv"), emit: csv
    path "versions.yml", emit: versions

    script:
    """
    set -euo pipefail
    imeta ls -C ${irodspath} \
        | grep -E 'attribute|value|units' \
        | sed -e 's/^attribute: //' -e 's/^value: //' -e 's/^units: //' \
        | awk 'NR%3!=0 {printf "\\\"%s\\\",", \$0} NR%3==0 {printf "\\\"%s\\\"\\n", \$0}' > irods_metadata.csv
    
    cat <<-END_VERSIONS > versions.yml
    "":
        irods: \$(ienv | grep version | awk '{ print \$3 }')
        awk: \$(awk --version | head -n1)
        sed: \$(sed --version | head -n1 | awk '{ print \$4 }')
    END_VERSIONS
    """

    stub:
    """
    touch irods_metadata.csv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        irods: \$(ienv | grep version | awk '{ print \$3 }')
        awk: \$(awk --version | head -n1)
        sed: \$(sed --version | head -n1 | awk '{ print \$4 }')
    END_VERSIONS
    """
}
