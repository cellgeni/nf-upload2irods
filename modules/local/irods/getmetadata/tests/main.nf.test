nextflow_process {
    name "Test Process: IRODS_GETMETADATA"
    script "../main.nf"
    process "IRODS_GETMETADATA"

    tag "modules"
    tag "irods"
    tag "irods/getmetadata"

    test("Get data object metadata") {
        tag "data_object"

        when {
            process {
                """
                input[0] = [ [id: "data_object_1"], "/seq/illumina/runs/41/41796/lane1/plex1/41796_1#1.cram" ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert snapshot(process.out).match() },
                { assert path(process.out.csv.get(0).get(1)).csv.columnCount == 3 }
            )
        }
    }

    test("Get collection metadata") {
        tag "collection"

        when {
            process {
                """
                input[0] = [ [id: "collection_1"], "/seq/illumina/cellranger-arc/cellranger-arc202_count_43c2d8dd1eaf98b635896165fd98ae3a" ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert snapshot(process.out).match() },
                { assert path(process.out.csv.get(0).get(1)).csv.columnCount == 3 }
            )
        }
    }

    test("Get linked collection metadata") {
        tag "collection"

        when {
            process {
                """
                input[0] = [ [id: "linked_collection_1"], "/archive/cellgeni/multiome/internal/cellranger_arc/6439/WS_wEMB13400246WS_wEMB13400220" ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert snapshot(process.out).match() },
                { assert path(process.out.csv.get(0).get(1)).csv.columnCount == 3 }
            )
        }
    }

    test("Get empty metadata") {
        tag "empty_metadata"

        when {
            process {
                """
                input[0] = [ [id: "emptymeta_collection_1"], "/archive/cellgeni/multiome" ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.versions },
                { assert process.out.csv },
                { assert snapshot(process.out).match() },
            )
        }
    }
}