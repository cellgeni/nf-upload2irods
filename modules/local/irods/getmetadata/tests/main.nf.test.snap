{
    "Get data object metadata": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "data_object_1"
                        },
                        "irods_metadata.csv:md5,197a5b8db4e7c8ae4783731992d6c6ac"
                    ]
                ],
                "1": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ],
                "csv": [
                    [
                        {
                            "id": "data_object_1"
                        },
                        "irods_metadata.csv:md5,197a5b8db4e7c8ae4783731992d6c6ac"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T12:14:27.465513263"
    },
    "Get linked collection metadata": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "linked_collection_1"
                        },
                        "irods_metadata.csv:md5,ce97a73816135627adc03848746fb881"
                    ]
                ],
                "1": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ],
                "csv": [
                    [
                        {
                            "id": "linked_collection_1"
                        },
                        "irods_metadata.csv:md5,ce97a73816135627adc03848746fb881"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T12:15:14.030409187"
    },
    "Get empty metadata": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "emptymeta_collection_1"
                        },
                        "irods_metadata.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ],
                "csv": [
                    [
                        {
                            "id": "emptymeta_collection_1"
                        },
                        "irods_metadata.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T13:12:45.149747057"
    },
    "Get collection metadata": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "collection_1"
                        },
                        "irods_metadata.csv:md5,fc54181e42872f3b1725271798b36dd4"
                    ]
                ],
                "1": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ],
                "csv": [
                    [
                        {
                            "id": "collection_1"
                        },
                        "irods_metadata.csv:md5,fc54181e42872f3b1725271798b36dd4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,aa918ef363c12d21e86ccdf876b0cb74"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-10-24T12:14:48.218053276"
    }
}